Add Mould King MK 5.0 advertising device

vicocz opened the pull request.

Commit:
Add MouldKing MK 5.0 advertisment device

J0EK3R commented:
Hey @vicocz  :)
...did you buy a MouldKing model including the MK5?

vicocz commented:
> Hey @vicocz  :)
> ...did you buy a MouldKing model including the MK5?

It's a gift and I need a reliable app to drive it :)

J0EK3R commented on BrickController2/BrickController2/DeviceManagement/MouldKing/MK5.cs:
```diff
@@ -0,0 +1,59 @@
+    /// <summary>
+    /// Base Telegram
+    /// </summary>
+    private static readonly byte[] Telegram_Base = [0x7d, 0x3f, 0x4d, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x82];
```
If the model is nibble-based and the zero value is 0x80/0x08 this should be 

```
private static readonly byte[] Telegram_Base = [0x7d, 0x3f, 0x4d, 0x88, 0x88, 0x88, 0x88, 0x88, 0x88, 0x82];
```

J0EK3R replied:
I'm just curious: how did you get the arrays for Telegram_Base and Telegram_Connect?

vicocz replied:
> If the model is nibble-based and the zero value is 0x80/0x08 this should be
> 
> ```
> private static readonly byte[] Telegram_Base = [0x7d, 0x3f, 0x4d, 0x88, 0x88, 0x88, 0x88, 0x88, 0x88, 0x82];
> ```

Probably, still testing...

vicocz replied:
> I'm just curious: how did you get the arrays for Telegram_Base and Telegram_Connect?

Still not sure if every byte is correct, but it works somehow. Still testing.

Tried simple hack and combinations: I've captured payload sent by MK IOS app and wrote unit test that incrementaly tries adding next byte from 0-255 range checking for expected prefix of encoded payload.

J0EK3R replied:
...I'm doing it the same way: incrementaly tries...
I'm asking myself if it would be possible to give Copilot the code for the encryption and ask it to create a decryption algorithm?

vicocz replied:
Btw are you aware of some disconnect datagram? MK5 device has some tank noise which is (obvisously) switched on on connection and it keeps doing such noise even though there is no datagram send for long period of time. And it's really annoying 😞

vicocz replied:
> ...I'm doing it the same way: incrementaly tries... I'm asking myself if it would be possible to give Copilot the code for the encryption and ask it to create a decryption algorithm?

Not sure about that, tried to use chatgpt to generate algorith for conversion of GUID to BLE 128bit byte array and it did not work properly. Even suggested an unit test which was not passing 😄

J0EK3R replied:
> Btw are you aware of some disconnect datagram? MK5 device has some tank noise which is (obvisously) switched on on connection and it keeps doing such noise even though there is no datagram send for long period of time. And it's really annoying

No, I didn't detect any disconnect datagrams (but I didn't look for them either).
I know the connect datagram to switch MK4/MK6 to BT mode and the control datagram.
How does controlling the MK5 work with the MK app?

J0EK3R replied:
> If the model is nibble-based and the zero value is 0x80/0x08 this should be
> 
> ```
> private static readonly byte[] Telegram_Base = [0x7d, 0x3f, 0x4d, 0x88, 0x88, 0x88, 0x88, 0x88, 0x88, 0x82];
> ```

...I was thinking about that: are you shure the MK5 controlling is nibble-based? Then there were up to 12 channels possible -> 6 * 0x88.
But your reverse-engineered control structure has 0x80 for the zero-value like a byte-based channel controlling...

vicocz replied:
I'm not, but the MK app does not allow to connect to different device as it does for MK6 and there are 3 axis/channels to control and one makro button.
Based on caught datagrams it sets first 2 bytes using 4bit per channel, which is nibble I guess. Have no idea why there are additional spaces.
Can chjeck it once more, e.g. on Adroid app, but the branch works I can control all motors as the original app.

J0EK3R replied:
If it works, everything's fine.
I was just wondering about the zero values of 0x80, which are 0x88 on the nibble-based MK4.

Commit:
renaming fix

Commit:
simplify device registration - 1st step

vicocz commented on BrickController2/BrickController2/UI/Controls/DeviceChannelSelector.xaml:
```diff
@@ -290,6 +290,28 @@
+                </Grid.ColumnDefinitions>
+
+                <Grid Grid.Column="0">
+                    <controls:ChannelSelectorRadioButton x:Name="MK5Channel3" DeviceType="MK5" Channel="2" HorizontalOptions="End" VerticalOptions="Start"/>
```
Fixed

Commit:
finetune UI - telegram

J0EK3R commented:
I also bought a tank model... the Tiger.

The engine noise is really annoying - I haven't found a way to turn it off yet.

Did you notice that you can also control the tank with the MK4 model?
So your approach with the Nible base class is definitely correct.

J0EK3R commented:
...please wait to merge this branch!

When examining the MK3.8/MK5 datagrams, I discovered that not all Nible-based models can be processed by the base class:
* The zero values ​​may differ from the 0x08 of the MK4.
* There are the familiar analog channels.
* There are channels for turning the motor left, right, and off.

I'm revising the base class.
---
I've written a dissector for wireshark to get the decrypted channel data, so I'm able to view the channel's changes of the MK App online...

J0EK3R commented:
I've revised MKBaseNibble to handle the "exotic" ;) MK3.8 and MK5.0 devices.

Here is the PR:

https://github.com/vicocz/brickcontroller2/pull/147

J0EK3R committed:
Revised MKBaseNibble and MK5 (#147)

* JK: revised MKBaseNibble and MK4.0
* JK: revised MK5

Commit:
nitpicks

J0EK3R commented:
...I have some ideas about handling the different channel types - I will revise the code again...

vicocz commented:
> ...I have some ideas about handling the different channel types - I will revise the code again...

OK, I plan to merge #148 first and adopt such API for registration.

Commit:
latest default branch

Commit:
revert unwanted changes

J0EK3R commented:
Here is the revision, a "type specific handling of channels"

https://github.com/vicocz/brickcontroller2/pull/149

J0EK3R commented:
Do you have assembled your model for testing or is just the motor on your desk?

vicocz commented:
> Here is the revision, a "type specific handling of channels"
> 
> https://github.com/vicocz/brickcontroller2/pull/149

I will have time for its review in one week time I guess.

vicocz commented:
> Do you have assembled your model for testing or is just the motor on your desk?

Just dev testing on my desky as it's a gift for somebody else.

J0EK3R commented:
> > Do you have assembled your model for testing or is just the motor on your desk?
> 
> Just dev testing on my desky as it's a gift for somebody else.

Here are my initial experiences:
Controlling the tank via the MouldKing remote control or our app is... hmmm... something special ;)
You can rotate the tank turret left and right (the remote control only has one speed; we have about four speed levels).
You can drive the tank forward and backward at different speeds—that's fine.
And you can turn left and right on the spot.
You can't drive curves by changing different speeds of the tank tracks.
When turning on the spot, the tank turret rotates in the opposite direction of travel and tries to maintain its direction. This doesn't work very well...

We could stop the tank turret from rotating on the spot by changing the zero value of the channel that controls the turret.

I noticed one small thing when controlling with our app:
If the joystick's zero position isn't exactly zero and so does not pass a zero value to method SetOutput, then the corresponding channel doesn't stop - I'll have to think about how to get this under control, a kind of zero-range.

J0EK3R commented:
For testing purposes, I added a fifth "virtual channel" to toggle the option to lock the turret when turning in place. This option can be enabled/disabled, for example, as a toggle channel mapped to a button on the PS controller.

Any interest? Should I push the code?

J0EK3R commented:
Here is what I've done:
I've developed a "type specific handling of channels including virtual channels"

https://github.com/vicocz/brickcontroller2/pull/149

J0EK3R commented:
The marked channel is an option to lock the turret when turning in place:
![Screenshot_2025-06-21-15-00-22-558-edit_com scn BrickController2](https://github.com/user-attachments/assets/6eb34e31-a3b9-4f1f-bca3-9d6294afe491)

![Screenshot_2025-06-21-15-01-28-752_com scn BrickController2](https://github.com/user-attachments/assets/8c4f1dad-e4ce-4c47-9188-6e821796004f)

[Tank.json](https://github.com/user-attachments/files/20846707/Tank.json)

J0EK3R commented:
The option channel 5 is defined as "SimpleToggle"

![Screenshot_2025-06-21-16-08-27-371-edit_com scn BrickController2](https://github.com/user-attachments/assets/9afc7ca3-5028-4413-8588-c97470d82515)

J0EK3R committed:
revised - type specific handling of channels (#149)

* JK: revised - channel types possible

* JK: added virtual channel handling

* JK: removed unnecessary call

* JK: fixed bad typo :|

* JK: revised SetOutput

* JK: fixed if condition

* JK: clean up

* JK: message of ArgumentException contains illegal value

* JK: copilot's suggestion

[nitpick] Variable 'realChannelNo' shadows the tuple element from setChannelHandler; consider deconstructing the tuple directly (e.g., 'var (realChannelNo, handler) = CreateChannelHandler(channelNo)') to avoid confusion.

* JK: lock to ensure ensure thread safety when updating the bitfield

* JK: reset option _turret_lock_Nibble on connect

* JK: reset _stored Values on InitDevice

* JK: switched to Vit's suggestion

* JK: implemented DisconnectDevice to enshure valid channel states in _bluetoothAdvertisingDeviceHandler

---------

Co-authored-by: J0EK3R <J0EK3R3.3@WIN11>

Commit:
Merge branch 'default' into local/mk5

Commit:
nitpicks from review

vicocz commented:
@J0EK3R THanx a lot for contribution to this feature.
